docs/generateDocs.py

```diff
@@ -0,0 +1,203 @@
+    or parses it from the arguments provided
+    """
+    global annotation_filter
+    if len(sys.argv) == 3:
+        f = sys.argv[2]
+        f = f.split(",")
+        for a in f:
+            annotation_filter.add(a)
+    else: 
+        annotation_filter = set(["when", "then", "given", "and", "but"])
+
+def print_help():
+    print("""
+USAGE: generateDocs.py DIRECTORIES [FILTER]
+
+Finds all cucumber steps in .java files in given directories annotated by given filter
+The final documentation is generated to docs/ directory
+Each \"module\" with at least one step defined is documented in its own file
+and every step is documented in index.md 
+
+Arguments:
+\tDIRECTORIES\t comma separated list of directories to find definitions in.
+\t\t (default value is the directory that the script is called from)
+\tFILTER\t comma separated list of annotations that define steps
+\t\t (default value is \"when,then,given,and,but\")
+
+Example usage: 
+    python3 generateDocs.py . "when,then"
+        """)
+
+def escape_regex(line):
+    """
+    Escapes the annotation line from the special characters used in markdown which could cause trouble
+    """
+    for c in "|>!#":
+        line = line.replace(c, "\\" + c)
+    return line
+
+def process_lines(file_name, lines, i):
+    """
+    Tries to find annotation on line with index i,
+    if an annotation is found then finds the function signature
+    """
+    annotation = lines[i].strip()
+    index = annotation.find("(")
+    #annotation was found
+    if (index != -1):
+        if annotation[1:index].lower() in annotation_filter:
+            #regex = @give(...)
+            annotation = escape_regex(annotation)
+            #looking for signature
+            signature = None
+            signature_incomplete = True
+            while signature_incomplete:
+                i += 1
+                line = lines[i].strip()
+                if line.startswith("public"):
+                    signature = line
+                    signature_incomplete = not signature.endswith("{")
+                elif signature:
+                    #Fixes incomplete signatures when they are longer than one line
+                    signature = signature + line
+                    signature_incomplete = not signature.endswith("{")
+            signature = signature.replace("{", "")
+            level = 1
+            starting_line_num = i
+            i += 1
+            steps.append(Annotation(file_name, starting_line_num, annotation, signature))
+            return i
+    return i
+
+def scan_file(file_name):
+    if not file_name.endswith(".java"):
+        return
+    try:
+        with open(file_name, "r") as f:
+            #Goes through all lines and tries to find annotations
+            line_num = 0
+            lines = f.readlines()
+            while line_num < len(lines):
+                line = lines[line_num].strip()
+                if line.startswith("@"):
```

Comment:
This can break if the annotation is commented using multiline comment:
```
/*
@When("why do we keep commented code? I don't know, but it can happen")
*/
```